### DISCUSSION TITLE
`iminuit` strategy error when using `cabinetry.fit.fit`

### DISCUSSION THREAD

**Discussion by TommyDESY** [scikit-hep/cabinetry#483] [Q&A] [answered]
Hi everyone,

I currently get an `iminuit` error using versions `0.6.0` of `cabinetry` and `0.7.6` of `pyhf` which had never occurred before. Below is the full error thread. `model` is built with `pyhf.pdf.Model`, `data` is a `numpy` array and `minos` is a list containing the name of my parameter of interest as a string (the error also occurs when the minos parameter is not set at all). I also tried to set the `iminuit` strategy manually in `cabinetry.fit.fit` to ensure that it's an integer but I get the same error. Does anyone know of a fix besides downgrading `iminuit` to an older version? Thanks for your help.

```
  File "/nfs/dust/belle2/user/tommy/Vub/inclusive_xulnu/xulnu/pyhf_utils.py", line 890, in extract_signal_strength_scale_factors
    fit_res = cabinetry.fit.fit(model, data, minos=minos)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nfs/dust/belle2/user/tommy/cabinetry/src/cabinetry/fit/__init__.py", line 367, in fit
    fit_results = _fit_model(
                  ^^^^^^^^^^^
  File "/nfs/dust/belle2/user/tommy/cabinetry/src/cabinetry/fit/__init__.py", line 159, in _fit_model
    result, corr_mat, best_twice_nll, result_obj = pyhf.infer.mle.fit(
                                                   ^^^^^^^^^^^^^^^^^^^
  File "/afs/desy.de/user/t/tommy/.local/lib/python3.11/site-packages/pyhf/infer/mle.py", line 131, in fit
    return opt.minimize(
           ^^^^^^^^^^^^^
  File "/afs/desy.de/user/t/tommy/.local/lib/python3.11/site-packages/pyhf/optimize/mixins.py", line 193, in minimize
    result = self._internal_minimize(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/afs/desy.de/user/t/tommy/.local/lib/python3.11/site-packages/pyhf/optimize/mixins.py", line 52, in _internal_minimize
    result = self._minimize(
             ^^^^^^^^^^^^^^^
  File "/afs/desy.de/user/t/tommy/.local/lib/python3.11/site-packages/pyhf/optimize/opt_minuit.py", line 119, in _minimize
    minimizer.strategy = strategy
    ^^^^^^^^^^^^^^^^^^
  File "/cvmfs/belle.cern.ch/el9/externals/v02-02-01/Linux_x86_64/common/lib/python3.11/site-packages/iminuit/minuit.py", line 228, in strategy
    self._strategy.strategy = value
    ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: (): incompatible function arguments. The following argument types are supported:
    1. (arg0: iminuit._core.MnStrategy, arg1: int) -> None

Invoked with: <iminuit._core.MnStrategy object at 0x7f08daedc830>, {}

**Comment by alexander-held** [comment]
Hi @TommyDESY, I have not seen this before and I am not aware of any particular issues with the latest `iminuit` version. Could you please share a way to reproduce this issue for me to have a closer look? Does changing the `iminuit` version change the behavior and can you pinpoint the version where the behavior changes?

**Reply by TommyDESY** [reply]
Hi @alexander-held,

here's a minimal example to reproduce the issue:

```
import pyhf
import cabinetry

spec_test = {
    'channels': [
        {'name': 'signal',
         'samples': [
             {'name': 'Xulnu_in',
              'data': [76, 102],
              'modifiers': [{'name': 'FEI_B0',
                             'type': 'normsys',
                             'data': {'hi': 0.9880450809276028, 'lo': 1.011954919072397}}]}]}]}

model = pyhf.Model(spec_test, poi_name='')
asimov_data = cabinetry.model_utils.asimov_data(model)
cabinetry.fit.fit(model, asimov_data)
```

I tested this small script with `iminuit` versions `2.15.1` to `2.26.0` and I always get the exact same error. Again, I use `pyhf v.0.7.6` and `cabinetry v0.6.0`.

**Reply by alexander-held** [reply]
I cannot reproduce the error with this. I think you probably have a modified local version of `cabinetry` (picked up from `/nfs/dust/belle2/user/tommy/cabinetry`) which differs from v0.6.0 and causes the difference. In the traceback
```pytb
  File "/nfs/dust/belle2/user/tommy/cabinetry/src/cabinetry/fit/__init__.py", line 159, in _fit_model
    result, corr_mat, best_twice_nll, result_obj = pyhf.infer.mle.fit(
                                                   ^^^^^^^^^^^^^^^^^^^
```
points to a line that is different from where this line is in the released version: https://github.com/scikit-hep/cabinetry/blob/2af877c58052c335572fcea4e604d49f32804e0a/src/cabinetry/fit/__init__.py#L94

Could you please see if any local modification you have done might be responsible for this error?

**Reply by TommyDESY** [reply, marked as answer]
Yes indeed, it was due to my changes to `cabinetry`, I should've noticed it. Thanks for the help!
